src/name.rs

```diff
@@ -16,126 +16,216 @@ use crate::{
     cert::{Cert, EndEntityOrCA},
     der, Error,
 };
+use core::fmt::Write as _;
 
-/// A DNS Name suitable for use in the TLS Server Name Indication (SNI)
-/// extension and/or for use as the reference hostname for which to verify a
-/// certificate.
+/// A reference to a DNS Name suitable for use in the TLS Server Name Indication
+/// (SNI) extension and/or for use as the reference hostname for which to verify
+/// a certificate.
 ///
-/// A `DnsName` is guaranteed to be syntactically valid. The validity rules are
-/// specified in [RFC 5280 Section 7.2], except that underscores are also
+/// A `DnsName` is guaranteed to be syntactically valid. The validity rules
+/// are specified in [RFC 5280 Section 7.2], except that underscores are also
 /// allowed.
 ///
-/// `DnsName` stores a copy of the input it was constructed from in a `String`
-/// and so it is only available when the `std` default feature is enabled.
+/// [RFC 5280 Section 7.2]: https://tools.ietf.org/html/rfc5280#section-7.2
+pub struct DnsName<B>(B)
+where
+    B: AsRef<[u8]>;
+
+/// A borrowed `DnsName`.
 ///
-/// `Eq`, `PartialEq`, etc. are not implemented because name comparison
-/// frequently should be done case-insensitively and/or with other caveats that
-/// depend on the specific circumstances in which the comparison is done.
+/// This is an alias for `DnsName<&'a [u8]>`.
+pub type DnsNameRef<'a> = DnsName<&'a [u8]>;
+
+/// An owned `DnsName`
 ///
-/// [RFC 5280 Section 7.2]: https://tools.ietf.org/html/rfc5280#section-7.2
+/// This is an alias for `DnsName<Box<[u8]>>`.
 #[cfg(feature = "std")]
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct DnsName(String);
+pub type DnsNameBox = DnsName<Box<[u8]>>;
```

Comment:
Coming back to this after a long time, I'm not sure why I thought it was important to make it especially convenient to support `Box<[u8]>`. I think I'll just remove this.